src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/SendPaymentRequests/SendPaymentRequestsCommandHandler.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using SFA.DAS.EmployerIncentives.Abstractions.Commands;
+using SFA.DAS.EmployerIncentives.Commands.Services.BusinessCentralApi;
+using SFA.DAS.EmployerIncentives.Data;
+using SFA.DAS.EmployerIncentives.Data.ApprenticeshipIncentives;
+
+namespace SFA.DAS.EmployerIncentives.Commands.ApprenticeshipIncentive.SendPaymentRequests
+{
+    public class SendPaymentRequestsCommandHandler : ICommandHandler<SendPaymentRequestsCommand>
+    {
+        private readonly IAccountDataRepository _accountRepository;
+        private readonly IPayableLegalEntityQueryRepository _queryRepository;
+        private readonly IBusinessCentralFinancePaymentsService _businessCentralFinancePaymentsService;
+
+        public SendPaymentRequestsCommandHandler(
+            IAccountDataRepository accountRepository,
+            IPayableLegalEntityQueryRepository queryRepository,
+            IBusinessCentralFinancePaymentsService businessCentralFinancePaymentsService)
+        {
+            _accountRepository = accountRepository;
+            _queryRepository = queryRepository;
+            _businessCentralFinancePaymentsService = businessCentralFinancePaymentsService;
+        }
+
+        public async Task Handle(SendPaymentRequestsCommand command, CancellationToken cancellationToken = default)
+        {
+            var payments = await _queryRepository.GetPaymentsToSendForAccountLegalEntity(command.AccountLegalEntityId);
+            if (payments.Any() == false)
+            {
+                return;
+            }
+
+            var sent = await _businessCentralFinancePaymentsService.SendPaymentRequestsForLegalEntity(payments);
+
+            await _accountRepository.UpdatePaidDateForPaymentIds(sent.PaymentsSent.Select(s => s.PaymentId).ToList(), command.AccountLegalEntityId, command.PaidDate);
+
+            if (!sent.AllPaymentsSent)
```

Comment:
Shouldn't this be handled in the IBusinessCentralFinancePaymentsService after all it's an implementation detail of that service that updates need to be paged.  So this handler should just call 

await _businessCentralFinancePaymentsService.SendPaymentRequestsForLegalEntity(payments); 
then 
 await _accountRepository.UpdatePaidDateForPaymentIds(sent.PaymentsSent.Select(s => s.PaymentId).ToList(), command.AccountLegalEntityId, command.PaidDate);

Reply:
The original intention was that this command handler would be called via a ServiceBus Message (which I think is still the longer term aim). And if the BC Api could not handle the number of payment requests, we would simply send the same SB message from within the handler. 

With the SB pattern, it meant we wouldn't spend a long time within the command handler, and we could be certain that the messages will eventually be sent to BC. It also meant that if BC was responding slowly or was down, our durable functions wouldn't start retrying the activity and sending even more requests to BC. 

However, mike asked that we didn't implement that yet. Hence this was the compromise pattern I put in place as it closely matches the original intent. I believe a subsequent story is on the backlog which might implement the SB message

Reply:
In that case can the filter to select the payments be done in the handler?  So move this line

 var paymentsToSend = payments.Take(_paymentRequestsLimit).ToList();

It just seems that the handler and the service have to work together to do the paged updates rather than just one being in control.  Would have to pass in the Options config into the handler to determine the limit

Reply:
good point, done